woocommerce/woocommerce-gutenberg-products-block
# Refactor external dispatch actions from being called inside useSelect

This PR resolves the long-standing issue in wordpress.com in which Cart and Checkout blocks would cause an infinite loop and, recently, crash some blocks.

The source of the issue was that the wpcom [code that did the monkey patching](https://github.com/Automattic/wp-calypso/blob/trunk/apps/wpcom-block-editor/src/wpcom/features/tracking.js#L767-L800) couldn't handle actions being called from inside a useSelect or useDispatch function, causing an infinite loop in Gutenberg.

I couldn't really pinpoint the exact line that handle that, maybe someone more experienced in that codebase can look cc @Addison-Stavlo.

For this PR, I just moved the logic from outside the useSelect call, it made sense anyway to do that outside.


Fixes https://github.com/woocommerce/woocommerce-blocks/issues/5866
Fixes https://github.com/Automattic/wp-calypso/issues/57449

### Note
Because we didn't make the wpcom code more resilient yet, this issue might surface back again, if it does, the fix would be to simply create your functions outside the useSelect function, not inside it.




### Testing



To test this PR, you will need to have a zip to upload to an atomic website `*.wpcomstaging.com`. Use this zip 
[woocommerce-gutenberg-products-block.zip](https://github.com/woocommerce/woocommerce-blocks/files/9144395/woocommerce-gutenberg-products-block.zip) that reflects this PR.

1. Create a new atomic website (wordpress.com -> new website -> pro plan -> install WooCommerce).
2. Upload and enable this plugin.
3. Open the editor, open the console as well.
4. Insert Cart block, you shouldn't see any console errors, all block should load fine without errors.
5. Insert Checkout block, you shouldn't see any console errors, all blocks should load fine without errors.
6. Try adding a product, going to cart and checkout, they should load fine.

Outside wpcom, make sure this PR doesn't regress anything:
- Create a coupon.
- Go to cart, try to use the coupon, you should see a snackbar below, no errors in console.
- Go to checkout, remove the coupon, insert it again, no error should surface.


* [ ] Do not include in the Testing Notes: Only include the second part of testing.

### WooCommerce Visibility





### Changelog

> Fix Cart an d Checkout blocks compatiblity issue with wordpress.com in which blocks wouldn't load in the editor.
